source/azure_iot_hub_client.c

```diff
@@ -626,3 +724,503 @@ AzureIoTHubClientResult_t AzureIoTHubClient_SendTelemetry( AzureIoTHubClient_t *
 
     return xResult;
 }
+/*-----------------------------------------------------------*/
+
+AzureIoTHubClientResult_t AzureIoTHubClient_ProcessLoop( AzureIoTHubClient_t * pxAzureIoTHubClient,
+                                                         uint32_t ulTimeoutMilliseconds )
+{
+    AzureIoTMQTTResult_t xMQTTResult;
+    AzureIoTHubClientResult_t xReturn;
+
+    if( pxAzureIoTHubClient == NULL )
+    {
+        AZLogError( ( "Failed to process loop: Invalid argument" ) );
+        xReturn = eAzureIoTHubClientInvalidArgument;
+    }
+    else if( ( xMQTTResult = AzureIoTMQTT_ProcessLoop( &( pxAzureIoTHubClient->_internal.xMQTTContext ),
+                                                       ulTimeoutMilliseconds ) ) != eAzureIoTMQTTSuccess )
+    {
+        AZLogError( ( "AzureIoTMQTT_ProcessLoop failed : ProcessLoopDuration=%u, Error=0x%08x",
+                      ulTimeoutMilliseconds, xMQTTResult ) );
+        xReturn = eAzureIoTHubClientFailed;
+    }
+    else
+    {
+        xReturn = eAzureIoTHubClientSuccess;
+    }
+
+    return xReturn;
+}
+/*-----------------------------------------------------------*/
+
+AzureIoTHubClientResult_t AzureIoTHubClient_SubscribeCloudToDeviceMessage( AzureIoTHubClient_t * pxAzureIoTHubClient,
+                                                                           AzureIoTHubClientCloudToDeviceMessageCallback_t xCallback,
+                                                                           void * prvCallbackContext,
+                                                                           uint32_t ulTimeoutMilliseconds )
+{
+    AzureIoTMQTTSubscribeInfo_t xMqttSubscription = { 0 };
+    AzureIoTMQTTResult_t xMQTTResult;
+    AzureIoTHubClientResult_t xReturn;
+    uint16_t usSubscribePacketIdentifier;
+    AzureIoTHubClientReceiveContext_t * pxContext;
+
+    if( ( pxAzureIoTHubClient == NULL ) ||
+        ( xCallback == NULL ) )
+    {
+        AZLogError( ( "Failed to subscribe cloud message : Invalid argument" ) );
+        xReturn = eAzureIoTHubClientInvalidArgument;
+    }
+    else
+    {
+        pxContext = &pxAzureIoTHubClient->_internal.xReceiveContext[ azureiothubRECEIVE_CONTEXT_INDEX_C2D ];
+        xMqttSubscription.xQoS = eAzureIoTMQTTQoS1;
```

Comment:
Dumb question and maybe answered in specs but why do we have Qos1 here but Qos0 on other subscriptions?

Reply:
Because C2D messages are subscribed with QoS1, where as all other are fire and forget type (Request/response)

[https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-mqtt-support#receiving-cloud-to-device-messages](https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-mqtt-support#receiving-cloud-to-device-messages)

When a device app subscribes to a topic with QoS 2, IoT Hub grants maximum QoS level 1 in the SUBACK packet. After that, IoT Hub delivers messages to the device using QoS 1.

---
In reply to: [612740875](https://github.com/Azure/azure-iot-middleware-freertos/pull/75#discussion_r612740875) [](ancestors = 612740875)